Guide.docc/IncrementalAdoption.md

```diff
@@ -1,3 +1,236 @@
+incrementally.
+
+The Swift 6 language mode must be explicitly enabled for existing projects.
+However, enabling the mode is just the first part of the migration process.
+It could take many changes, across different modules to complete.
+To make this practical, these changes can be made _incrementally_.
+This allows you to progress without disrupting the entire project.
+
+Swift includes a number of language features and standard library APIs to help
+make incremental adoption easier.
+
+## Wrapping Callback-Based Functions
+
+APIs that accept and invoke a single function on completion are an extremely
+common pattern in Swift.
+It's possible to make a version of such a function that is usable directly from 
+an asynchronous context.
+
+```swift
+func updateStyle(backgroundColor: ColorComponents, completionHandler: @escaping () -> Void) {
+    // ...
+}
+```
+
+This is an example of a function that informs a client its work is complete
+using a callback.
+You can wrap this function up into an asynchronous version using
+_continuations_.
+
+```swift
+func updateStyle(backgroundColor: ColorComponents) async {
+    withCheckedContinuation { continuation in
+        updateStyle(backgroundColor: backgroundColor) {
+            continuation.resume()
+        }
+    }
+}
+```
+
+The `withCheckedContinuation` function is one of a [suite of standard library
+APIs][continuation-apis] that exist to make interfacing non-async and async code possible.
+
+> Note: Introducing asynchronous code into a project can surface data isolation
+checking violations. To understand and address these, see [Crossing Isolation Boundaries][]
+
+[Crossing Isolation Boundaries]: commonproblems#Crossing-Isolation-Boundaries
+[continuation-apis]: https://developer.apple.com/documentation/swift/concurrency#continuations
+
+## Dynamic Isolation
+
+Expressing the isolation of your program statically, using annotations and
+other language constructs, is both powerful and concise.
+But, because static isolation uses Swift's type system, introducing it
+can potentially require changes to all places a type is used.
+
+Dynamic isolation provides runtime mechanisms for describing data isolation.
+These are a more manual, but also more flexible alternative.
+Dynamic isolation can be an essential tool for interfacing a Swift 6 component
+with another that has not yet migrated,
+even if these components are within the _same_ module.
+
+### Internal-Only Isolation
+
+Suppose you have determined that a reference type within your project can be
+best described with `MainActor` static isolation.
+
+```swift
+@MainActor
+class WindowStyler {
+    private var backgroundColor: ColorComponents
+
+    func applyStyle() {
+        // ...
+    }
+}
+```
+
+This `MainActor` isolation may be _logically_ correct.
+But, if this type is used in other unmigrated locations,
+adding static isolation here could require many additional changes.
```

Comment:
I think if this is the case, you should use `@preconcurrency @MainActor` to stage in the diagnostics as warnings. `MainActor.assumeIsolated` is useful when you _cannot_ write `@MainActor` on a type because the isolation is controlled by an API that you don't own, like a nonisolated protocol requirement or superclass method that you're overriding from some dependency that isn't yours.

Reply:
I have another section on `@preconcurrency @MainActor`. Do you think I should just remove this? I think some people might appreciate a path that lets them move forward in a sub-optimal-but-warning-free way.

Reply:
Hmm. Let's merge this as is and we can continue to talk about this section!